### ISSUE TITLE
new rule: if a function returns a promise, then calling code must do something with it

### ISSUE THREAD

**Issue by sethobrien** [Microsoft/tslint-microsoft-contrib#174]
I think this would have to wait until palantir/tslint#680 so we have better information about the type of functions, but I'll suggest it anyway.

As an example, consider the following:

```
interface ISomething {
    getNumber(): number;
    isActive(): boolean;
}
function getStringRepresentation(something: ISomething): string {
    return something.isActive() ? something.getNumber().toString() : "";
}
```

If somebody changes `getNumber` or `isActive` to be async (and their team's coding conventions don't say to append `Async` to all async methods), then `getStringRepresentation` would still compile fine but no longer accomplish what it expected.

Our proposed solution is that any time you call a function that returns a `Promise`, you must do one of the following:
- Call `.catch`
- Call `.then`
- Use `await` to get the result
- Store the `Promise` someplace other than a local variable

Any `Promise` returned by `.then` and `.catch` would be exempt, to avoid requiring an infinite series of `.then` calls.

**Comment by HamletDRC** [comment]
thanks @sethobrien for this idea. I am tagging it "feature-request", "blocked on type checker", and "microsoft-internal". 

You are correct in that the type checker limitation currently makes this near impossible to implement. 

Is this a real bug that you encountered on your team?

**Comment by sethobrien** [comment]
Not me personally, but @JoshuaKGoldberg did.

**Comment by JoshuaKGoldberg** [comment]
Can confirm. In https://github.com/JoshuaKGoldberg/intern-assassins/, for example...

```typescript
if (this.isTrue()) {
    // ...
}
```

...would work fine if `isTrue` is non-async, but once converted to return a Promise will always be truthy.

**Comment by HamletDRC** [comment]
the tslint project just pushed code to support the improved type checker.

**Comment by JoshuaKGoldberg** [comment]
@HamletDRC is there an issue to track including the improved type checker? I've prototyped out this rule and would love to flesh it out and send a PR.

https://github.com/JoshuaKGoldberg/bellevuejs-linting/blob/master/5.%20Advanced%20TSLint%20Rule/rules/noFloatingPromisesRule.ts

**Comment by HamletDRC** [comment]
Hi @JoshuaKGoldberg , thanks for the patience while I was on vacation! 

All of the issues that have to do with the improved type checker are labeled "blocked on type checker"

I think you are taking the right approach. I would just write a whole bunch of unit tests and make sure it really does what you think.

**Comment by JoshuaKGoldberg** [comment]
Welcome back :)

**Comment by JoshuaKGoldberg** [comment]
@sethobrien looks like this issue can be closed?

**Comment by wwwouter** [comment]
Hi, 

I'm use the no-floating-promises rule, but this code mentioned here

```
if (this.isTrue()) {
    // ...
}
```

doesn't seem to get caught by that rule.

Did any of you create another rule for this case?

**Comment by wwwouter** [comment]
Just ignore this, I created my own rule for this specific case (https://www.npmjs.com/package/@wwwouter/tslint-contrib)

**Comment by JoshuaKGoldberg** [comment]
@wwwouter It should be caught by `strict-boolean-expressions` too, if that helps!

**Comment by wwwouter** [comment]
@JoshuaKGoldberg Thanks for your response!

I tried strict-boolean-expressions, but even after some fine-tuning there were more than 200 errors in our current code base. Next time when I start a new project, I'll sure use that rule from the get-go!
